lib/record_store/provider/ns1.rb

```diff
@@ -0,0 +1,222 @@
+require_relative 'ns1/client'
+
+module RecordStore
+  class Provider::NS1 < Provider
+    class Error < StandardError; end
+
+    class << self
+      def client
+        Provider::NS1::Client.new(api_key: secrets['api_key'])
+      end
+
+      # Downloads all the records from the provider.
+      #
+      # Returns: an array of `Record` for each record in the provider's zone
+      def retrieve_current_records(zone:, stdout: $stdout)
+        full_api_records = records_for_zone(zone).map do |short_record|
+          client.record(
+            zone: zone,
+            fqdn: short_record["domain"],
+            type: short_record["type"],
+            must_exist: true,
+          )
+        end
+
+        full_api_records.map { |r| build_from_api(r, zone) }.flatten
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        client.zones.map do |zone|
+          zone["zone"]
+        end
+      end
+
+      private
+
+      # Fetches simplified records for the provided zone
+      def records_for_zone(zone)
+        client.zone(zone)["records"]
+      end
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        new_answers = [{ answer: build_api_answer_from_record(record) }]
+
+        record_fqdn = record.fqdn.sub(/.$/, '')
```

Comment:
This will remove any character at the end of the string ? Was it supposed to match the dot (`.`) ? In which case, `\.` would be better, or even nicer to avoid regex for nothing:
```suggestion
        record_fqdn = record.fqdn.delete_suffix('.')
```